### PR TITLE
Introduce transaction for precompiled letters

### PR DESCRIPTION
**Pull request by servingUpAces**
Wrap the saving the notification and uploading the precompiled letter to s3 in a transaction.

If the upload to s3 fails the notification will not be saved to the
database.

### CODE REVIEW

File: tests/app/v2/notifications/test_post_letter_notifications.py [hunk 1]
```diff
@@ -673,6 +673,28 @@ def test_post_precompiled_letter_notification_returns_201(
...
+        "content": "bGV0dGVyLWNvbnRlbnQ="
+    }
+
+    auth_header = create_authorization_header(service_id=sample_service.id)
+    with pytest.raises(expected_exception=Exception):
+        client.post(
+            path="v2/notifications/letter",
+            data=json.dumps(data),
+            headers=[('Content-Type', 'application/json'), auth_header])
+
```
Comment (idavidmcdonald) [inline tests/app/v2/notifications/test_post_letter_notifications.py:693]: Optional: I think what would make this test even more rigerous would be for us to assert that `create_letter_notification` was called. So then it's very clear that we called `create_letter_notification`, then an exception was thrown but at the end there was no notification saved to the DB

**Review by idavidmcdonald — APPROVED** [review]
Nice one.

As part of my review, I ran the unit test you wrote having removed the `with_transaction` statement you had added and it does indeed break so leaves me confident that this is doing what it says it does

File: tests/app/v2/notifications/test_post_letter_notifications.py (hunk 1 above)
Comment (servingUpAces) [reply tests/app/v2/notifications/test_post_letter_notifications.py:693]: I've never mocked a method, then used side_effect=method... neat

**Comment by leohemsted** [comment]
hmm im surprised this works! given that within the transaction, `dao_create_notification` also has the `@transactional` wrapper - guess postgres transactions didn't work like i expected them to. 🤔

but if the tests pass... 🤷‍♂️ ✅

**Comment by idavidmcdonald** [comment]
I was confused by this too but then found out it was because I needed to pull master and get new changes which change it to use `autocommit` rather than `transactional`.

See https://github.com/alphagov/notifications-api/blob/91542ad33eb995c9b3e617cde30a9714846dd37b/app/dao/notifications_dao.py#L82

**Pull request merged**